File: backend/src/entities/Attendance.ts
import { Entity, PrimaryGeneratedColumn, Column, ManyToOne, JoinColumn, Unique } from 'typeorm';
import { Student } from './Student';

@Entity('attendance')
@Unique(['student_id', 'date'])
export class Attendance {
  @PrimaryGeneratedColumn()
  id: number;

  @Column()
  student_id: number;

  @Column({ type: 'date' })
  date: Date;

  @Column({ 
    type: 'enum', 
    enum: ['לא דיווח', 'נוכח', 'יצא', 'יום לא בא לי', 'חיסור מאושר', 'אישור היעדרות קבוע'],
    default: 'לא דיווח'
  })
  status: 'לא דיווח' | 'נוכח' | 'יצא' | 'יום לא בא לי' | 'חיסור מאושר' | 'אישור היעדרות קבוע';

  @Column({ 
    type: 'enum', 
    enum: ['ללא', 'איחור', 'נסגר אוטומטית'],
    default: 'ללא'
  })
  sub_status: 'ללא' | 'איחור' | 'נסגר אוטומטית';

  @Column({ type: 'enum', enum: ['student', 'manager', 'auto'] })
  reported_by: 'student' | 'manager' | 'auto';

  @Column({ type: 'time', nullable: true })
  check_in_time: string;

  @Column({ type: 'time', nullable: true })
  check_out_time: string;

  @Column({ 
    type: 'enum', 
    enum: ['n/a', 'manual', 'auto_16'],
    default: 'n/a'
  })
  closed_reason: 'n/a' | 'manual' | 'auto_16';

  @Column({ default: false })
  override_locked: boolean;

  @Column({ type: 'timestamp', nullable: true })
  override_locked_at: Date;

  @ManyToOne(() => Student, student => student.attendances)
  @JoinColumn({ name: 'student_id' })
  student: Student;
}

File: backend/app/models/attendance.py
from sqlalchemy import Column, Integer, String, Date, DateTime, Boolean, ForeignKey, Enum, UniqueConstraint
from sqlalchemy.orm import relationship
from .base import Base
import enum
from datetime import datetime

class AttendanceStatus(enum.Enum):
    NOT_REPORTED = "לא דיווח"
    PRESENT = "נוכח"
    LEFT = "יצא"
    DIDNT_FEEL_LIKE_IT = "יום לא בא לי"
    APPROVED_ABSENCE = "חיסור מאושר"
    PERMANENT_ABSENCE_APPROVAL = "אישור היעדרות קבוע"

class SubStatus(enum.Enum):
    NONE = "ללא"
    LATE = "איחור"
    AUTO_CLOSED = "נסגר אוטומטית"

class ReportedBy(enum.Enum):
    STUDENT = "student"
    MANAGER = "manager"
    AUTO = "auto"

class ClosedReason(enum.Enum):
    NA = "n/a"
    MANUAL = "manual"
    AUTO_16 = "auto_16"

class Attendance(Base):
    __tablename__ = "attendance"
    
    id = Column(Integer, primary_key=True, index=True)
    student_id = Column(Integer, ForeignKey("students.id"), nullable=False)
    date = Column(Date, nullable=False, index=True)
    status = Column(Enum(AttendanceStatus), nullable=False, default=AttendanceStatus.NOT_REPORTED)
    sub_status = Column(Enum(SubStatus), nullable=False, default=SubStatus.NONE)
    reported_by = Column(Enum(ReportedBy), nullable=False, default=ReportedBy.STUDENT)
    check_in_time = Column(DateTime, nullable=True)
    check_out_time = Column(DateTime, nullable=True)
    closed_reason = Column(Enum(ClosedReason), nullable=False, default=ClosedReason.NA)
    override_locked = Column(Boolean, nullable=False, default=False)
    override_locked_at = Column(DateTime, nullable=True)
    
    # Relationships
    student = relationship("Student", back_populates="attendances")
    
    # Constraints
    __table_args__ = (
        UniqueConstraint('student_id', 'date', name='unique_student_date'),
    )